File: packages/snow_draw_core/lib/draw/types/element_geometry.dart
import 'package:meta/meta.dart';

import 'draw_point.dart';
import 'draw_rect.dart';

/// Geometry snapshot used by move operations.
@immutable
class ElementMoveSnapshot {
  const ElementMoveSnapshot({required this.center});
  final DrawPoint center;
}

/// Geometry snapshot used by resize operations.
@immutable
class ElementResizeSnapshot {
  const ElementResizeSnapshot({required this.rect, required this.rotation});
  final DrawRect rect;
  final double rotation;

  DrawPoint get center => rect.center;
  double get width => rect.width;
  double get height => rect.height;
}

/// Geometry snapshot used by rotate operations.
@immutable
class ElementRotateSnapshot {
  const ElementRotateSnapshot({required this.center, required this.rotation});
  final DrawPoint center;
  final double rotation;
}

File: packages/snow_draw_rust_engine/src/draw/types/element_geometry.rs
#![allow(dead_code)]

use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;
use serde::{Deserialize, Serialize};

/// Geometry snapshot used by move operations.
#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
pub struct ElementMoveSnapshot {
    pub center: DrawPoint,
}

/// Geometry snapshot used by resize operations.
#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
pub struct ElementResizeSnapshot {
    pub rect: DrawRect,
    pub rotation: f64,
}

impl ElementResizeSnapshot {
    pub fn center(self) -> DrawPoint {
        self.rect.center()
    }

    pub fn width(self) -> f64 {
        self.rect.width()
    }

    pub fn height(self) -> f64 {
        self.rect.height()
    }
}

/// Geometry snapshot used by rotate operations.
#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
pub struct ElementRotateSnapshot {
    pub center: DrawPoint,
    pub rotation: f64,
}
